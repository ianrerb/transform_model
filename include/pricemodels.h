#ifndef PRICEMODELS_H
#define PRICEMODELS_H
#define PI 3.14159265358979
#include<complex.h>
#include<fftw3.h>
#include<math.h>
#include<vector>

struct Option {double strike; double premium; }; //struct to hold strike, premium pairs

class fft_base {//abc
  public:
    //ctor and dtor
    fft_base(){ };
    virtual ~fft_base(){ };
    
    //defined functions
    double getLambda(const unsigned int &N, const double &eta){ return 2.0*PI/(static_cast<double>(N)*eta);};
    std::vector<Option> PutPrices(const double &Spot,
				  const double &Strike, 
				  const double &C, 
				  const unsigned int &N, 
				  const double &alpha, 
				  const double &eta);
    
    //pure virtual functions
    virtual double complex logCF(const double &S, const double complex &u) const=0;
  
  private:
    //helper function
    void ComputeX(double complex X[], 
		  const double &Spot,
		  const double &Strike, 
		  const double &C, 
		  const unsigned int &N, 
		  const double &alpha, 
		  const double &eta); 
  };

class GBM_model : public fft_base {
  public:
    GBM_model(const double sigma_, const double r_, const double q_, const double t_): sigma(sigma_), r(r_), q(q_), t(t_) { }; 
    ~GBM_model(){ }; 
    double complex logCF(const double &S, const double complex &u) const;
  private:
    double sigma, r, q, t;
  };


#endif
