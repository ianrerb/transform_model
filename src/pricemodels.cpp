#include"../include/pricemodels.h"

void fft_base::ComputeX(double complex X[], const double &S, const double &C, const double &offset){
  double s = log(S);
  double lambda = getLambda();

  for(int j = 0; j!=N; j++){
    double complex nu = j*static_cast<double complex>(N); //nu = (j-1)eta
    double complex val1 = alpha+I*nu; //denominator of C term
    double complex temp = eta*C/(val1*(1.0+val1))*cexp(-I*(s-PI/eta))*logCF(S,nu-(alpha+1)*I); //x_j j!=0
    if(j==0){ temp*=.5; }//adjustment for j = 0
    X[j]=temp;
    }
  }

std::vector<double> fft_base::PutPrices(const double &S, const double &C, const double &offset){
  double complex xvals[N];
  printf("Computing X...\n"); 
  fft_base::ComputeX(xvals,S,C,offset);
  double complex yvals[N];
  fftw_plan p = fftw_plan_dft_1d(N,xvals,yvals,FFTW_FORWARD, FFTW_ESTIMATE);
  std::vector<double> retvals;
  printf("Computing FFT...\n"); 
  fftw_execute(p);
//  fftw_destroy_plan(p);
  for(int i =0; i!= N; i++)
    retvals.push_back(exp(-alpha*(log(S)-static_cast<double>(N)*getLambda()/2.0))*creal(yvals[i])/PI);
  return retvals;
  
  }

double complex GBM_model::logCF(const double &S, const double complex &u) const{
  double complex retval = I*(clog(S) + (r - q - cpow(sigma,2.0)/2.0)*t)*u - cpow(sigma*u,2.0)*t/2.0;
  return cexp(retval);
  }
