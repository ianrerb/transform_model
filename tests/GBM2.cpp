/* Testing with Values in Computational Methods in Finance Chapter 2, page 42 (section2.1.5)  */ 

#include"../include/pricemodels.h"
#include<cmath>
#include<iostream>
#include<vector>

using namespace std;

int main(){
  double Spot = 100;
  double Strike = 90;
  GBM_model gbm(.3,.05,0,1);
  
  vector<double> alpha;
  alpha.push_back(0.01);  
  alpha.push_back(0.5);  
  alpha.push_back(1);  
  alpha.push_back(1.5);  
  alpha.push_back(2);  
  alpha.push_back(5);  
  alpha.push_back(10);  
 
  vector<double> eta;
  eta.push_back(0.15); 
  eta.push_back(0.1);
  eta.push_back(0.05); 

  vector<unsigned int> N;
  N.push_back(pow(2,6));
  N.push_back(pow(2,8));
  N.push_back(pow(2,10));
 
  for(vector<double>::iterator a_iter=alpha.begin(); a_iter!= alpha.end(); a_iter++){
    for(vector<double>::iterator e_iter=eta.begin(); e_iter!= eta.end(); e_iter++){
      for(vector<unsigned int>::iterator N_iter = N.begin(); N_iter!=N.end(); N_iter++){
	vector<Option> results =gbm.PutPrices(Spot,Strike,exp(-.05),*N_iter,*a_iter,*e_iter);
	cout<<results[*N_iter/2].premium<<" ";
	} 
      }
      cout<<endl;
    }
  
  return 0;
  }
