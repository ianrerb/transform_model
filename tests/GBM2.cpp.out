-157.793 -157.793 -157.793 -29.9503 -29.9503 -29.9503 -14.9751 -14.9751 -14.9751 
-2.25138 -2.25138 -2.25138 -0.427328 -0.427328 -0.427328 -0.213664 -0.213664 -0.213664 
-0.915729 -0.915729 -0.915729 -0.173812 -0.173812 -0.173812 -0.0869058 -0.0869058 -0.0869058 
-0.541781 -0.541781 -0.541781 -0.102834 -0.102834 -0.102834 -0.0514169 -0.0514169 -0.0514169 
-0.384179 -0.384179 -0.384179 -0.0729198 -0.0729198 -0.0729198 -0.0364599 -0.0364599 -0.0364599 
-0.262873 -0.262873 -0.262873 -0.0498951 -0.0498951 -0.0498951 -0.0249475 -0.0249475 -0.0249475 
-3.36905 -3.36905 -3.36905 -0.63947 -0.63947 -0.63947 -0.319735 -0.319735 -0.319735 
